docs/rfcs/2023-03-29-file-external-table/rfc.md

```diff
@@ -0,0 +1,83 @@
+---
+Feature Name: "File external table"
+Tracking Issue: https://github.com/GreptimeTeam/greptimedb/issues/1041
+Date: 2023-03-08
+Author: "Xu Wenkang <wenymedia@gmail.com>"
+---
+
+File external table
+
+---
+
+# Summary
+
+Allows users to perform SQL queries on files
+
+# Motivation
+
+User data may already exist in other storages, i.g., file systems/s3, etc. in CSV, parquet, JSON format, etc. We can provide users the ability to perform SQL queries on these files.
+
+
+# Details
+## Overview
+![overview](external-table-engine-overview.png)
+
+We implement a file table engine that creates an external table by accepting user-specified file paths and treating all files as immutable.
+
+1. File table manifest: contains file paths and connection options.
+2. File table engine: provides the ability to decode files to `RecordBatch` stream.
+
+## Life cycle
+
+### Register table
+1. Writes table metadata to manifest
+2. Creates the table via file table engine
+3. Registers table to catalog
+### Deregister table
+1. Fetches the target table (figure out table engine type)
+2. Deregisters the target table
+3. Finds the target table engine
+4. Drops the target table
+
+```rust 
+    engine_manager.engine("file")?.drop_table("target")
+```
+### Recover
+1. Collects tables name and engine type info
+2. Finds the target tables in different engines
+3. Opens and registers table
+
+## Drawbacks
+- Unsupported filter pushdown
+- Hard to support indexing
+
+# Alternatives
+
+## Using datafusion API
+We can use datafusion API to register a file table:
+```rust
+let ctx = SessionContext::new();
+
+ctx.register_csv("example", "tests/data/example.csv", CsvReadOptions::new()).await?;
+
+// create a plan
+let df = ctx.sql("SELECT a, MIN(b) FROM example WHERE a <= b GROUP BY a LIMIT 100").await?;
+```
+### Drawbacks
+The datafusion implements its own `Object Store` abstraction, which makes it impossible to use our `CacheLayer`.
```

Comment:
Can you elaborate the `CacheLayer` a little more?

Reply:
> Can you elaborate the `CacheLayer` a little more?

Yes, I will, in fact, It's just the `LruCacheLayer`, we used in the `OpenDAL`